From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>
Date: Tue, 10 Dec 2024 13:54:15 +0100
Subject: [PATCH 16/29] TODO Add a test to show the race condition

TODO: list the managed resources
TODO: how to simulate that the secret is used, but the managed resource is not yet updated.

This commit adds a unit test that shows that the Garbage Collector
of the Gardener Resource Manager has a race condition.

It uses the Cooperative Executor that was introduced in the previous commits.

The Cooperative Executor is used to run the Garbage Collector and an actor that
reconciles a managed resource and a secret, with the managed resource
referencing the secret. The Garbage Collector and this actor are executed
concurrently. The Kubernetes client API calls are interleaved in all possible
ways.

In the initial state of the system there is one unreferenced secret. The unit
test shows that there is an ordering of the API calls where the Garbage
Collector deletes the secret that is "reused" by the actor. Neither the Garbage
Collector, nor the actor detects an error, and hence the system is left in an
inconsistent state indefinitely.

Why "reused"? Immutable secrets have content based names. If the actor
reconciles a secret with the same content, it will reuse already existing secret
with the same name.

The purpose of this PR is to prevent the "silent" deletion of such secrets by
the Garbage Collector. The fix is to detect conflicts between the Garbage
Collector's delete method call and the client "reusing" the secret.

The unit test shows the 3 orderings of the relevant API calls:

Path 1:
- GC/0 List secrets
- GC/1 Delete unused secret
- Actor/0 Get secret --> not found
- Actor/1 Create secret

In this path, the secret is deleted by the Garbage Collector and recreated by
the actor.

Path 2:
- Actor/0 Get secret --> the immutable secret exists
                         no need to create or update
- GC/0 List secrets
- GC/1 Delete unused secret

TODO

Path 3 is similar to Path 2:
- GC/0 List secrets
- Actor/0 Get secret --> the immutable secret exists
                         no need to create or update
- GC/1 Delete unused secret

The GC deletes the secret, although it is referenced in the meantime.
---
 .../garbagecollector/reconciler_test.go       | 176 ++++++++++++++++++
 1 file changed, 176 insertions(+)

diff --git a/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go b/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go
index f14d5dbafc..5092e3a805 100644
--- a/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go
+++ b/pkg/resourcemanager/controller/garbagecollector/reconciler_test.go
@@ -6,8 +6,11 @@ package garbagecollector_test
 
 import (
 	"context"
+	"fmt"
+	"strings"
 	"time"
 
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
 	"github.com/go-logr/logr"
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
@@ -19,6 +22,7 @@ import (
 	testclock "k8s.io/utils/clock/testing"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
@@ -27,6 +31,7 @@ import (
 	"github.com/gardener/gardener/pkg/logger"
 	"github.com/gardener/gardener/pkg/resourcemanager/apis/config"
 	. "github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector"
+	"github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/race"
 	"github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/references"
 )
 
@@ -47,6 +52,7 @@ var _ = Describe("Collector", func() {
 	BeforeEach(func() {
 		log = logger.MustNewZapLogger(logger.DebugLevel, logger.FormatText)
 		logf.SetLogger(log.WithName("garbagecollector"))
+		log = log.WithName("test            ")
 		c = fakeclient.NewClientBuilder().WithScheme(kubernetes.SeedScheme).Build()
 		fakeClock = testclock.NewFakeClock(creationTimestamp.Add(minimumObjectLifetime / 2))
 		gc = &Reconciler{
@@ -57,6 +63,176 @@ var _ = Describe("Collector", func() {
 		}
 	})
 
+	Describe("#raceCondition", func() {
+		var (
+			executor     race.CooperativeExecutor
+			setupClients func()
+			clientGC     client.Client
+			clientActor  client.Client
+		)
+		BeforeEach(func() {
+			executor = race.NewCooperativeExecutor(race.INFO)
+			setupClients = func() {
+				clientGC = race.NewCooperativeClient("GC", c, executor)
+				gc.TargetClient = clientGC
+				clientActor = race.NewCooperativeClient("Actor", c, executor)
+			}
+		})
+
+		When("there is an unreferenced secret in the system", func() {
+			var (
+				secret      *corev1.Secret
+				setupSecret func()
+			)
+			BeforeEach(func() {
+				setupSecret = func() {
+					secret = &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+						Name:      "secret",
+						Namespace: metav1.NamespaceDefault,
+					}}
+					kubernetesutils.MakeUnique(secret)
+					Expect(client.IgnoreNotFound(c.Delete(ctx, secret))).To(Succeed())
+					Expect(c.Create(ctx, secret)).To(Succeed())
+				}
+			})
+			When("an actor reconciles a MR referencing that secret", func() {
+				var (
+					cleanupMR          func()
+					assertSecretExists func() error
+				)
+				BeforeEach(func() {
+					cleanupMR = func() {
+						Expect(client.IgnoreNotFound(c.Delete(ctx, &resourcesv1alpha1.ManagedResource{ObjectMeta: objectMetaFor("mr")}))).To(Succeed())
+					}
+				})
+
+				It("the referenced secret should exist at the end", func() {
+					assertSecretExists = func() error {
+						secretList := &corev1.SecretList{}
+						Expect(c.List(ctx, secretList)).To(Succeed())
+						if len(secretList.Items) != 1 {
+							return fmt.Errorf("a (concurrently) referenced secret was deleted by the Garbage Collector")
+						}
+						return nil
+					}
+					setup := func() {
+						setupClients()
+						setupSecret()
+						cleanupMR()
+						fakeClock.SetTime(creationTimestamp.Add(minimumObjectLifetime / 2))
+					}
+					assert := assertSecretExists
+					GC := func() {
+						runs := 0
+						for {
+							result, err := gc.Reconcile(ctx, reconcile.Request{})
+							log.Info(fmt.Sprintf("GC Reconcile() result: %+v, err: %v", result, err))
+							runs++
+							if result.Requeue && result.RequeueAfter < gc.Config.SyncPeriod.Duration {
+								fakeClock.Step(result.RequeueAfter)
+							} else {
+								if err == nil {
+									break
+								}
+							}
+						}
+						log.Info(fmt.Sprintf("GC finished after %d runs", runs))
+					}
+					actor := func() {
+						for {
+							secret1 := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+								Name:      "secret",
+								Namespace: metav1.NamespaceDefault,
+							}}
+							kubernetesutils.MakeUnique(secret1)
+
+							mr := &resourcesv1alpha1.ManagedResource{ObjectMeta: objectMetaFor("mr", secret1)}
+							result, err := controllerutil.CreateOrUpdate(ctx, clientActor, mr, func() error { return nil })
+							log.Info(fmt.Sprintf("CreateOrUpdate managed resource returned: OperationResult=%+v, error=%v", result, err))
+							if err != nil {
+								continue
+							}
+
+							secret2 := secret1.DeepCopy()
+							mutateSecret := func() error {
+								secret2.Labels = secret1.Labels
+								secret2.Immutable = secret1.Immutable
+								return nil
+							}
+
+							result, err = controllerutil.CreateOrUpdate(ctx, clientActor, secret2, mutateSecret)
+							log.Info(fmt.Sprintf("CreateOrUpdate secret           returned: OperationResult=%+v, error=%v", result, err))
+							if err == nil {
+								break
+							}
+						}
+					}
+					result := executor.RunAllCombinations(setup, assert, GC, actor)
+					Expect(result.NumberOfPathsChecked).To(Equal(8))
+					Expect(strings.TrimSpace(result.PathsChecked)).To(Equal(strings.TrimSpace(strings.ReplaceAll(`
+
+  Path 1:
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, ]}, error=<nil>
+    assertion error: <nil>
+  Path 2:
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, ]}, error=<nil>
+    assertion error: <nil>
+  Path 3:
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, ]}, error=<nil>
+    assertion error: <nil>
+  Path 4:
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: []}, error=<nil>
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+  - GC/2 Delete(&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, []) -> error=<nil>
+    assertion error: a (concurrently) referenced secret was deleted by the Garbage Collector
+  Path 5:
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, ]}, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+    assertion error: <nil>
+  Path 6:
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, ]}, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+    assertion error: <nil>
+  Path 7:
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: []}, error=<nil>
+  - GC/2 Delete(&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, []) -> error=<nil>
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=secrets "secret-e3b0c442" not found
+  - Actor/2 Create(&Secret{Namespace: default, Name: secret-e3b0c442, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+    assertion error: <nil>
+  Path 8:
+  - GC/0 List(SecretList, [map[resources.gardener.cloud/garbage-collectable-reference:true]]) -> list=&PartialObjectMetadataList{Items: [&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, ]}, error=<nil>
+  - GC/1 List(ManagedResourceList, []) -> list=&PartialObjectMetadataList{Items: []}, error=<nil>
+  - Actor/0 Create(&ManagedResource{Namespace: default, Name: mr, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, []) -> object=&ManagedResource{Namespace: default, Name: mr, ResourceVersion: 1, Annotations: map[reference.resources.gardener.cloud/secret-ccbd158d:secret-e3b0c442], }, error=<nil>
+  - GC/2 Delete(&PartialObjectMetadata{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], }, []) -> error=<nil>
+  - Actor/1 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=secrets "secret-e3b0c442" not found
+  - Actor/2 Create(&Secret{Namespace: default, Name: secret-e3b0c442, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true], Immutable: true, }, error=<nil>
+    assertion error: <nil>
+
+`, "\n  ", "\n"))), "Here is the actual value of PathsChecked in full length. Copy this and update the string literal in the unit test's source code: \n\n%s\n",
+						result.PathsChecked)
+					Expect(result.NumberOfPathsWithFailedAssertion).To(Equal(1))
+				})
+			})
+		})
+	})
+
 	Describe("#collectGarbage", func() {
 		var (
 			unlabeledSecret    *corev1.Secret
